json: not encoding embedded struct

### Describe the bug

Try to encode a json to a struct with another embedded.

### Reproduction Steps

```v
module main

import json

struct Json2 {
	inner []f64
}

struct Json {
	Json2
	test f64
}

fn main() {
	str:= '{"inner": [1, 2, 3, 4, 5],"test": 1.0}'
	data := json.decode(Json, str) or {
		eprintln('Failed to decode json, error: ${err}')
		return
	}
	println(data)
	assert data.inner.len == 5
	assert data.inner[0] == 1.0
	assert data.inner[4] == 5.0
	assert data.test == 1.0

	data_json := json.encode(data)
	assert data_json == str
}
```

### Expected Behavior

Run correctly, with all of the data encoded

### Current Behavior

```
Json{
    Json2: Json2{
        inner: [1.0, 2.0, 3.0, 4.0, 5.0]
    }
    test: 1.0
}
break.v:27: FAIL: fn main.main: assert data_json == str
   left value: data_json = {"test":1}
  right value: str = {"inner": [1, 2, 3, 4, 5],"test": 1.0}
V panic: Assertion failed...
v hash: 5e624a5
/tmp/v_1000/break.tmp.c:7496: at _v_panic: Backtrace
/tmp/v_1000/break.tmp.c:14048: by main__main
/tmp/v_1000/break.tmp.c:14088: by main
make: *** [Makefile:11: run] Error 1
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V full version: V 0.4.7 5e624a5

### Environment details (OS name and version, etc.)

V full version: V 0.4.7 5e624a5
OS: linux, Ubuntu 24.04.1 LTS
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 7840HS w/ Radeon 780M Graphics

vexe: /home/esquerbatua/git/v/v
vexe mtime: 2024-09-21 15:56:29

vroot: OK, value: /home/esquerbatua/git/v
VMODULES: OK, value: /home/esquerbatua/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.43.0
Git vroot status: weekly.2024.37-35-g5e624a5b
.git/config present: true

CC version: cc (Ubuntu 13.2.0-23ubuntu4) 13.2.0
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/json/json_embed_test.v
+++ b/vlib/json/json_embed_test.v
@@ -1,0 +1,31 @@
+import json
+
+struct Json3 {
+	embed f64
+}
+
+struct Json2 {
+	Json3
+	inner []f64
+}
+
+struct Json {
+	Json2
+	test f64
+}
+
+fn test_main() {
+	str := '{"inner": [1, 2, 3, 4, 5],"test": 1.0, "embed": 2.0}'
+	data := json.decode(Json, str) or {
+		eprintln('Failed to decode json, error: ${err}')
+		return
+	}
+	println(data)
+	assert data.inner.len == 5
+	assert data.inner[0] == 1.0
+	assert data.inner[4] == 5.0
+	assert data.test == 1.0
+	assert data.embed == 2.0
+
+	assert dump(json.encode(data)) == '{"embed":2,"inner":[1,2,3,4,5],"test":1}'
+}

--- a/vlib/json/json_encode_embed_test.v
+++ b/vlib/json/json_encode_embed_test.v
@@ -1,0 +1,27 @@
+import json
+
+struct Json2 {
+	inner []f64
+}
+
+struct Json {
+	Json2
+	test f64
+}
+
+fn test_main() {
+	str := '{"inner":[1,2,3,4,5],"test":1.2}'
+	data := json.decode(Json, str) or {
+		eprintln('Failed to decode json, error: ${err}')
+		return
+	}
+	println(data)
+	assert data.inner.len == 5
+	assert data.inner[0] == 1.0
+	assert data.inner[4] == 5.0
+	assert data.test == 1.2
+
+	data_json := json.encode(data)
+	dump(data_json)
+	assert data_json == str
+}

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -614,6 +614,7 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 		mut is_skip := false
 		mut is_required := false
 		mut is_omit_empty := false
+		mut skip_embed := false
 
 		for attr in field.attrs {
 			match attr.name {
@@ -768,10 +769,10 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 						if embed == int(field.typ) {
 							g.gen_struct_enc_dec(field.typ, g.table.sym(field.typ).info,
 								styp, mut enc, mut dec, name)
+							skip_embed = true
 							break
 						}
 					}
-					continue
 				}
 				tmp := g.new_tmp_var()
 				gen_js_get_opt(dec_name, field_type, styp, tmp, name, mut dec, is_required)
@@ -797,15 +798,15 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 				dec.writeln('\t}')
 			}
 		}
-		if embed_prefix.len > 0 {
-			return
+		if skip_embed {
+			continue
 		}
 		// Encoding
 		mut enc_name := js_enc_name(field_type)
 		prefix_enc := if utyp.has_flag(.option) {
-			'(*(${g.base_type(utyp)}*)val.data)'
+			'(*(${g.base_type(utyp)}*)val${embed_member}.data)'
 		} else {
-			'val'
+			'val${embed_member}'
 		}
 		is_option := field.typ.has_flag(.option)
 		indent := if is_option { '\t\t' } else { '\t' }

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -767,8 +767,13 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 				if name.len > 0 && name[0].is_capital() && field_sym.info is ast.Struct {
 					for embed in info.embeds {
 						if embed == int(field.typ) {
+							prefix_embed := if embed_prefix != '' {
+								'${embed_prefix}.${name}'
+							} else {
+								name
+							}
 							g.gen_struct_enc_dec(field.typ, g.table.sym(field.typ).info,
-								styp, mut enc, mut dec, name)
+								styp, mut enc, mut dec, prefix_embed)
 							skip_embed = true
 							break
 						}

